File: Project_ngan/Frontend/book-app/src/types/book.ts
export type SortKey =
  | "TITLE_ASC"
  | "TITLE_DESC"
  | "YEAR_ASC"
  | "YEAR_DESC"
  | "CREATED_DESC";

export type Book = {
  id: number;
  title: string;
  author: string;
  publishYear?: number | null;
  tags?: string | null; // "khoa hoc,ky nang"
  pdfUrl?: string | null;
  hasLocalFile: boolean;
  createdAt?: string | null;
  updatedAt?: string | null;
};

export type BookUpsert = {
  title: string;
  author: string;
  publishYear?: number | null;
  tags?: string | null;
  pdfUrl?: string | null;
  hasLocalFile: boolean;
};

File: Project_ngan/Backend/spring/spring/src/main/java/com/newwave/spring/entity/Book.java
package com.newwave.spring.entity;

import jakarta.persistence.*;
import lombok.Data;
import lombok.RequiredArgsConstructor;
import org.hibernate.annotations.Nationalized;

import java.time.LocalDateTime;

@Data
@RequiredArgsConstructor
@Entity
@Table(name = "books")
public class Book {

    @Id
    @GeneratedValue(strategy = GenerationType.IDENTITY)
    private Long id;

    @Nationalized
    @Column(nullable = false)
    private String title;

    @Nationalized
    @Column(nullable = false)
    private String author;

    private Integer publishYear;

    @Nationalized
    @Column(length = 500)
    private String tags;

    @Column(length = 500)
    private String pdfUrl;

    @Column(nullable = false)
    private boolean hasLocalFile;

    private LocalDateTime createdAt;
    private LocalDateTime updatedAt;

    @PrePersist
    protected void onCreate() {
        createdAt = LocalDateTime.now();
        updatedAt = createdAt;
    }

    @PreUpdate
    protected void onUpdate() {
        updatedAt = LocalDateTime.now();
    }

    // getters & setters
}

